kpi/models/object_permission.py

```diff
@@ -948,3 +962,194 @@ def _update_partial_permissions(self, user_id, perm, remove=False,
+                   'permission_id',
+                   'permission__codename',
+                   'deny')
+        object_permissions_per_user = defaultdict(list)
+        for record in records:
+            object_permissions_per_user[record['user_id']].append((
+                record['permission_id'],
+                record['permission__codename'],
+                record['deny'],
+            ))
+
+        return object_permissions_per_user
+
+    @staticmethod
+    @cache_for_request
+    def __get_all_user_permissions(content_type_id, user_id):
+        """
+        Retrieves all object permissions and build an dict with object ids as keys.
+        Useful to retrieve permissions (thanks to `@cache_for_request`)
+        for several objects in a row without fetching data from data again & again
+
+        Because `django_cache_request` creates its keys based on method's arguments,
+        it's important to minimize its number to hit the cache as much as possible.
+        This method should be called when object permissions for a specific user
+        are needed several times in a row (within the same request).
+
+        It will hit the DB once for this user. If object permissions are needed
+        for an another object (i.e. `Asset`, `Collection`), in subsequent calls,
+        they can be easily retrieve by the returned dict keys.
+
+        Args:
+            content_type_id (int): ContentType's pk
+            user_id (int): User's pk
+
+        Returns:
+            dict: {
+                '<object_id>': [
+                    (permission_id, permission_codename, deny),
+                    (permission_id, permission_codename, deny),
+                    ...
+                ],
+                '<object_id>': [
+                    (permission_id, permission_codename, deny),
+                    (permission_id, permission_codename, deny),
+                    ...
+                ]
+            }
+        """
+        records = ObjectPermission.objects. \
+            filter(content_type_id=content_type_id, user=user_id).\
+            values('object_id',
+                   'permission_id',
+                   'permission__codename',
+                   'deny')
+
+        object_permissions_per_object = defaultdict(list)
+        for record in records:
+            object_permissions_per_object[record['object_id']].append((
+                record['permission_id'],
+                record['permission__codename'],
+                record['deny'],
+            ))
+
+        return object_permissions_per_object
+
+    def __get_object_permissions(self, is_denied, user=None, codename=None):
+        """
+        Returns a set of user ids and object permission ids related to
+        object `self`.
+
+        Args:
+            is_denied (bool): If `True`, returns denied permissions
+            user (User)
+            codename (str)
+
+        Returns:
+            set: [(User's pk, Permission's pk)]
```

Comment:
I'm in favor to use `.pk` everywhere we need to use a `Django` object's primary key.  (which I started to do since I read https://stackoverflow.com/questions/2165865/django-queries-id-vs-pk)
Even if I would keep `user_id` as a variable (or argument).

Reply:
Great, thanks for investigating